File: solutions/haskell/perfect-numbers/5/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

divisors :: Int -> [Int]
divisors n = filter (\m -> n `mod` m == 0) [1 .. n `div` 2]

aliquotSum :: Int -> Int
aliquotSum n = sum (divisors n)

classify :: Int -> Maybe Classification
classify n
 | n <= 0    = Nothing
 | otherwise = Just (pickClass n (aliquotSum n))
 where
   pickClass n aliquot
     | aliquot == n = Perfect
     | aliquot < n  = Deficient
     | aliquot > n  = Abundant

File: solutions/erlang/perfect-numbers/2/src/perfect_numbers.erl
-module(perfect_numbers).
-export([classify/1]).

classify(N) when N =< 0 -> erlang:error("Input is not positive integer");
classify(N) ->
  Divisors = [X || X <- lists:seq(1, N div 2), N rem X == 0],
  nicomachusClass(lists:sum(Divisors), N).

nicomachusClass(AliquotSum, AliquotSum) -> perfect;
nicomachusClass(AliquotSum, N) when AliquotSum > N -> abundant;
nicomachusClass(AliquotSum, N) when AliquotSum < N -> deficient.
